reconcile/openshift_resources.py

```diff
@@ -0,0 +1,277 @@
+                    cluster,
+                    namespace,
+                    resource_type,
+                    openshift_resource.name,
+                    openshift_resource
+                )
+
+        # Desired State
+        openshift_resources = namespace_info.get('openshiftResources') or []
+        for resource in openshift_resources:
+            openshift_resource = None
+
+            provider = resource['provider']
+            if provider == 'resource':
+                path = resource['path']
+                try:
+                    openshift_resource = fetch_provider_resource(path)
+                except FetchResourceError as e:
+                    error("provider resource. path: {}. ".format(path)
+                          + str(e.message))
+                    continue
+            else:
+                error('Unknown provider: {}'.format(provider))
+                continue
+
+            if openshift_resource is None:
+                error('Invalid None type for openshift_resource.')
+                continue
+
+            # add to inventory
+            try:
+                ri.add_desired(
+                    cluster,
+                    namespace,
+                    openshift_resource.kind,
+                    openshift_resource.name,
+                    openshift_resource
+                )
+            except KeyError:
+                # This is failing because in the managed_type loop (where the
+                # `initialize_resource_type` method was called), this specific
+                # combination was not initialized, meaning that it shouldn't be
+                # managed. But someone is trying to add it via app-interface
+                error("Unknown cluster/namespace/kind: {}/{}/{}.".format(
+                    cluster,
+                    namespace,
+                    openshift_resource.kind,
+                ))
+                continue
+
+    return oc_map, ri, errors
+
+
+def apply(dry_run, oc_map, c, n, rt, item):
+    logging.info(['apply', c, n, rt, item.name])
+
+    if not dry_run:
+        item.annotate()
+        body = OpenshiftResource.serialize(item.body)
+        oc_map[c].apply(n, body)
+
+
+def delete(dry_run, oc_map, c, n, rt, name):
+    logging.info(['delete', c, n, rt, name])
+
+    if not dry_run:
+        oc_map[c].delete(n, rt, name)
+
+
+def run(dry_run=False):
+    gqlapi = gql.get_api()
+
+    namespaces_query = gqlapi.query(NAMESPACES_QUERY)['namespaces']
+
+    oc_map, ri, errors = fetch_data(namespaces_query)
+
+    for c, n, rt, data in ri.iterate():
+        # desired items
+        for name, d_item in data['desired'].items():
+            c_item = data['current'].get(name)
+
+            if c_item is not None:
+                if c_item.has_qontract_annotations():
+                    if c_item.sha256sum() == d_item.sha256sum():
+                        # don't apply if sha256sum hashes match
+                        continue
```

Comment:
would it make sense to add DEBUG log here which can help to trouble shoot things?

Reply:
:+1: